lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js

```diff
@@ -0,0 +1,142 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var resolve = require( 'path' ).resolve;
+var tape = require( 'tape' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var abs = require( '@stdlib/math/base/special/abs' );
+var PI_SQUARED = require( '@stdlib/constants/float64/pi-squared' );
+var EPS = require( '@stdlib/constants/float64/eps' );
+var tryRequire = require( '@stdlib/utils/try-require' );
+
+
+// FIXTURES //
+
+var small = require( './fixtures/python/small.json' );
+var medium = require( './fixtures/python/medium.json' );
+var large = require( './fixtures/python/large.json' );
+
+
+// VARIABLES //
+
+var spence = tryRequire( resolve( __dirname, './../lib/native.js' ) );
+var opts = {
+	'skip': ( spence instanceof Error )
+};
+
+
+// TESTS //
+
+tape( 'main export is a function', opts, function test( t ) {
+	t.ok( true, __filename );
+	t.strictEqual( typeof spence, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function accurately computes the dilogarithm for small positive numbers', opts, function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var v;
+	var i;
+
+	x = small.x;
+	expected = small.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		v = spence( x[ i ] );
+		delta = abs( v - expected[ i ] );
+		tol = 2.0 * EPS * abs( expected[ i ] );
+		t.ok( delta <= tol, 'within tolerance. x: ' + x[ i ] + '. Value: ' + v + '. Expected: ' + expected[ i ] + '. Tolerance: ' + tol + '.' );
+	}
+	t.end();
+});
+
+tape( 'the function accurately computes the dilogarithm for medium positive numbers', opts, function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var v;
+	var i;
+
+	x = medium.x;
+	expected = medium.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		v = spence( x[ i ] );
+		delta = abs( v - expected[ i ] );
+		tol = 8.3 * EPS * abs( expected[ i ] );
```

Comment:
This test tolerance differs from the JavaScript tolerance. Is there a reason why given the implementation?

Reply:
This was the reason why I mentioned about different tolerances required for `C` and `javascript` implementations. I checked the `C` implementation, both against the original `cephes` and our `javascript` implementations, but still couldn't find any difference.

Reply:
@gunjjoshi It looks like the primary potential culprit is `ln`. I suggest manually checking to see there whether results diverge for medium sized input values to Spence.

Reply:
In `ln`, there is a rounding off happening for the `C` results, unlike that of `javascript` ones.

For instance,
In `C`, `ln(0.3261523046092184)` gives `-1.120391`, whereas in `javascript`, it comes out to be `-1.12039081466284`.
Similarly, 
In `C`, `ln(0.8436498997995991)` gives `-0.170018`, whereas in `javascript`, it comes out to be `-0.17001768110790474`.

Should this be a thing of concern?

Reply:
@gunjjoshi We would want the C implementation to match the JavaScript one. In general, we want the implementations to yield equivalent results. In this case, results shouldn't be rounded / truncated in the `ln` C implementation. Would be great if you were able to track that down and fix in a separate PR! 🙏

Reply:
Make sure that, when you are logging C values, you specify the appropriate precision in `printf` statements. That rounding looks suspicious to me and not the result of `ln`, per se.

Reply:
When I printed the results in `C` with `%.ng`, where `n` is the number of digits to be displayed after the decimal point, I found that the results do match the `javascript` ones. So, the problem might be somewhere else, as `ln` yields the same results for both `C` and `javascript` implementations, for medium sized input values.

Reply:
@kgryte @Planeshifter The problem seems to be at [L137](https://github.com/stdlib-js/stdlib/pull/2298/files#diff-762c1402388106d6bc0fe6a7c2e39048be27d1972dcf6c623b120c9747654f7dR137). 

Both `C` and `javascript` produce different results. For instance,

When `x = 0.9558294589178357`, 
`C`: `polyval_a( w ) / polyval_b( w ): 1.0112649603618284`
`JavaScript`: `polyvalA( w ) / polyvalB( w ): 1.0112649603618282`

Observe the difference in the last digits here.

Should this be due to the difference in the internal division implementation in `C` and `javascript`, leading to a little difference in the precision? In that case we might need different tolerances.

Reply:
@gunjjoshi Thanks for tracking this down! Interesting that division would yield differences. It's also possible that the polynomial equations are slightly different.

Reply:
Okay. I tracked down the issue. The discrepancy arises from floating-point operation fusing (see https://clang.llvm.org/docs/UsersManual.html#cmdoption-ffp-contract). If we disable fusing, we seem to get expected results.

In short, it is fine for us to have a different tolerance for this package, as there isn't much we can do to actually ensure the same results in the wild, so to speak. Depends on what compiler flags are set, rather than something to do with our implementation.

Reply:
Got it. So how do we detect this floating-point operation fusing? Like, in our future implementations of other functions too, this might arise, right?

Reply:
In this case, we can see the difference by setting the tolerance back to `2.0 * EPS` and then doing

```
$ CFLAGS="-ffp-contract=off" make install-node-addons NODE_ADDONS_PATTERN="math/base/special/spence"
$ make test TESTS_FILTER=".*/math/base/special/spence/.*"
```

and then

```
$ CFLAGS="-ffp-contract=on" make install-node-addons NODE_ADDONS_PATTERN="math/base/special/spence"
$ make test TESTS_FILTER=".*/math/base/special/spence/.*"
```

This has probably happened in the past; we just hadn't caught it until now. Realistically, it is unlikely to be a problem. Main thing is that we investigated and ruled out implementation bugs. So long as we are able to explain it, should be fine.

Reply:
I added a comment to the `test.native.js` file. I think we can consider this discussion resolved.

Reply:
Great, thanks for the detailed explanation!